### PR TITLE
fix ailing integration tests in test_public_api

### PR DESCRIPTION
**Pull request by a-dubs**
Fixes: GH #428 :tada: 

---

### EC2 minimal images:
manually verified my changes by modifying and running `examples/ec2.py`:
```diff
diff --git a/examples/ec2.py b/examples/ec2.py
index 84f025e..9a362c4 100755
--- a/examples/ec2.py
+++ b/examples/ec2.py
@@ -137,24 +137,15 @@ def demo():
     through a number of examples.
     """
     with pycloudlib.EC2(tag="examples") as ec2:
-        key_name = "test-ec2"
-        handle_ssh_key(ec2, key_name)
-
-        daily = ec2.daily_image(release="bionic")
-        daily_pro = ec2.daily_image(release="bionic", image_type=ImageType.PRO)
-        daily_pro_fips = ec2.daily_image(
-            release="bionic", image_type=ImageType.PRO_FIPS
-        )
-
-        launch_basic(ec2, daily)
-        launch_pro(ec2, daily_pro)
-        launch_pro_fips(ec2, daily_pro_fips)
-        custom_vpc(ec2, daily)
-        snapshot(ec2, daily)
-        launch_multiple(ec2, daily)
-        hot_add(ec2, daily)
+        from pprint import pprint
+        pprint(ec2.find_matching_images("minimal")[0])
+        pprint(ec2.find_matching_images("ubuntu*server")[0])
+        pprint(ec2.released_image(
+            release="noble",
+            image_type=ImageType.MINIMAL,
+        ))
 
 
 if __name__ == "__main__":
-    logging.basicConfig(level=logging.DEBUG)
+    logging.basicConfig(level=logging.INFO)
     demo()
diff --git a/pycloudlib/ec2/cloud.py b/pycloudlib/ec2/cloud.py
index fc57e40..ba5ea48 100644
--- a/pycloudlib/ec2/cloud.py
+++ b/pycloudlib/ec2/cloud.py
@@ -194,6 +194,37 @@ class EC2(BaseCloud):
                 "Values": [arch],
             },
         ]
+    
+    # function that will find all images matching a substring
+    def find_matching_images(
+        self,
+        substring: str,
+        include_deprecated: bool = False,
+    ):
+        filters = [
+            {
+                "Name": "name",
+                "Values": [
+                    f"*{substring}*"
+                ],
+            },
+        ]
+        owner = self._get_owner(image_type=ImageType.GENERIC)
+
+        images = self.client.describe_images(
+            Owners=[owner],
+            Filters=filters,
+            IncludeDeprecated=include_deprecated,
+        )
+
+        if not images.get("Images"):
+            raise ValueError(
+                "Could not find image matching {}".format(
+                    substring
+                )
+            )
+
+        return images["Images"]
 
     def _find_latest_image(
         self,

```

**THEN** run the ec2 public api integration tests:
```
tox -e integration-tests -- tests/integration_tests/test_public_api.py -k "ec2"
```

### Testing of GCE service account issues
Ran the following locally (I have service account) and @mstpn ran locally on his machine:
```bash
tox -e integration-tests -- tests/integration_tests/test_public_api.py -k "gce""
```
This runs both normal and minimal public api integration tests for GCE.

### Testing of SSH key failed re-uploading
This is tested by running ec2 public api integrations as mentioned earlier.

### CODE REVIEW

**Comment by mstpn** [comment]
Tested this locally and both offending `test_public_api` tests from #428 now pass.
```
tox -e integration-tests -- tests/integration_tests/test_public_api.py::test_public_api -k "gce"
tox -e integration-tests -- tests/integration_tests/test_public_api.py::test_public_api_minimal_images -k "gce"
```

**Comment by a-dubs** [comment]
@blackboxsw @mstpn The holy trinity has landed. Feast your eyes

**Comment by a-dubs** [comment]
also: **PLEASE** rebase when merging - all 3 commits are separate logical commits

**Comment by mstpn** [comment]
LGTM. Thanks for working through this with me and helping resolve something I missed in #409.

**Comment by a-dubs** [comment]
okay finally properly ready for review @blackboxsw

File: pycloudlib/ec2/cloud.py [hunk 1]
```diff
@@ -139,20 +139,24 @@ def _get_name_for_image_type(
...
+                disk_type,
             )
             if release in LTS_RELEASES:
-                return "{}/ubuntu-{}{}-*-server{}-*".format(
+                return "{}/ubuntu-{}{}-*-{}-*".format(
                     base_location,
                     release,
                     "-daily" if daily else "",
-                    "-minimal" if image_type == ImageType.MINIMAL else "",
+                    "minimal" if image_type == ImageType.MINIMAL else "server",
```
Comment (blackboxsw) [inline pycloudlib/ec2/cloud.py:152]: Thank you for this fix, this explains the inability to find minimal images on my PR. thx

Comment (a-dubs) [reply pycloudlib/ec2/cloud.py:152]: Yessir 😈😤💯🔥

**Review by blackboxsw — APPROVED** [review]
Ship it! Ran through minimal ec2 integration tests for focal, jammy, noble and oracular ensuring we can obtain minimal images appropriately. Validated GCP as well continues to work for service_account_email config in both pycloudlib.toml and GCE ~/.config/gcloud/*creds.json

**Comment by a-dubs** [comment]
@blackboxsw thank you so much for thoroughly retesting this!

**Comment by a-dubs** [comment]
> LGTM, tox formatting/linting notwithstanding. Thanks for working through this with me and helping resolve something I missed in #409.

My pleasure! Thanks for helping out with this :)

**Pull request merged**